### PR TITLE
MSTR-90: Add charsRemaining helper to UI Core

### PR DESCRIPTION
**Pull request by masmerino13**

### CODE REVIEW

**Comment by masmerino13** [comment]
Yeah, jQuery validation achive something similar, BUT just to validate, this helper is to display inline the number of `remaining characters` (this isn't a validation helper)

In AppEx we have some pages where this helper is needed and looks like a nice to have in the UI Core.

You can take a look at the screenshot below (I already sent you the Abstract mockup by DM also)
helper

@azefiel

**Comment by azefiel** [comment]
Nice, that does sound like a great addition.

Have you looked for existing jQuery plugins instead of reinventing the wheel? I would assume there is bunch to choose from as this feature seems pretty common.

**Comment by masmerino13** [comment]
Yeah, I think so :)

No, to be honest, I doesn't take a look for plugins, BUT yeah, this is a pretty common and useful feature :)

**Comment by masmerino13** [comment]
Hi @azefiel 

Do you think that I should consider to use a plugin intead of something custom?

This script is pretty simple, but just let me know so I can looks for a plugin or if it's ok for you let's merge the current stuff :)

Thank you!

**Comment by azefiel** [comment]
Hey, apologies @masmerino13 I somehow didn't receive notifications from this thread.

I am not against using your implementation of the feature. However, the following requirements should be taken care of before this PR get merged:

- use a more descriptive function name: something like `charsRemaining` would make more sense
- add a documentation page ([example](https://docs.2600hz.com/ui/docs/monster-ui/disableAutoFill()/))
- add jsdoc function header ([example](https://github.com/2600hz/monster-ui/blob/master/src/js/lib/monster.ui.js#L3360))
- add explicit error handling ([example](https://github.com/2600hz/monster-ui/blob/master/src/js/lib/monster.ui.js#L3367))

**Comment by masmerino13** [comment]
Thanks @azefiel,

I will let you know when those changes makes ready :)

**Comment by masmerino13** [comment]
Hi @azefiel 

I did the changes, any comments/suggestions are welcome!

Thanks!

**Comment by azefiel** [comment]
The CircleCI check is failing, if you do not have access to the pipelines dashboard, here is the error:

> page  docs/monster-ui/charsRemaining().md  is not valid

**Comment by masmerino13** [comment]
Weird, I have duplicated another file ;(

**Comment by masmerino13** [comment]
Hi @azefiel

I don't found what's wrong with my doc file, also it's curoius to me why the test was success for the PR in branch 4.3 (https://github.com/2600hz/monster-ui/pull/479)

Is there a way to validate `.md` files?

**Comment by masmerino13** [comment]
Finally!!! The file name should include `().md`

File: src/js/lib/monster.ui.js [hunk 1]
```diff
@@ -3698,6 +3698,61 @@ define(function(require) {
...
+			checkLength = function(event) {
+				var $target = $(event.target);
+
+				currentLenght = type === 'div' ? $target[0].textContent.length : $target.val().length;
+				allowedChars = size - currentLenght;
+
+				if (allowedChars <= 0 || allowedChars > size) {
+					label.addClass('maxlength-error');
+
+					event.preventDefault();
```
[the remark this answers is not in the document]
Comment (masmerino13) [reply src/js/lib/monster.ui.js]: yes

**Comment by masmerino13** [comment]
Hi @azefiel ,

I did some changes in my PR, it will be nice (when have a chance) if you take a look a this.

Thanks

File: src/js/lib/monster.ui.js [hunk 2]
```diff
@@ -3698,6 +3698,61 @@ define(function(require) {
...
+
+		if (!_.isUndefined(args) && !_.isPlainObject(args)) {
+			throw TypeError('"options" is not a plain object');
+		}
+
+		var $this = $($target),
+			size = args.size,
+			customClass = args.customClass || '',
+			type = args.type,
+			currentLenght = $this.val().length,
```
[the remark this answers is not in the document]
Comment (masmerino13) [reply src/js/lib/monster.ui.js]: yes, this was tested with `wysiwyg` editor also, seems on this case the prop has `contenteditable="inherit"`

btw, if it is a div the I looks for his content, why we needs to check for 'contenteditable=true'?

File: docs/monster-ui/charsRemaining().md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+## Syntax
+```javascript
+monster.ui.charsRemaining($target, args);
+```
+
+### Parameters
+Key | Description | Type | Default | Required
+:-: | --- | :-: | :-: | :-:
+`$target` | Form element to be checked. | `jQuery` | | `true`
+`args` | Aguments to setup the helper | `Object` | | `true`
```
Comment (azefiel) [inline docs/monster-ui/charsRemaining().md]:
```suggestion
`args` | Aguments to setup the helper | `Object`([#/args](#args)) | | `true`
```
(from [example](https://raw.githubusercontent.com/2600hz/monster-ui/master/docs/monster-ui/monthpicker().md))

File: docs/monster-ui/charsRemaining().md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+### args
```
Comment (azefiel) [inline docs/monster-ui/charsRemaining().md]:
```suggestion
#### `args`
```
(from [example](https://raw.githubusercontent.com/2600hz/monster-ui/master/docs/monster-ui/monthpicker().md))

File: docs/monster-ui/charsRemaining().md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+Key | Description | Type | Default | Required
+:-: | --- | :-: | :-: | :-:
+`args.size` | The maxlength to be validated | `Integer` | | `true`
```
Comment (azefiel) [inline docs/monster-ui/charsRemaining().md]:
```suggestion
`size` | The maxlength to be validated | `Integer` | | `true`
```
(from [example](https://raw.githubusercontent.com/2600hz/monster-ui/master/docs/monster-ui/monthpicker().md))

Also, as stated in my previous review, [there is not such thing as a pure integer type in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Data_types); your should use `Number` instead.

File: docs/monster-ui/charsRemaining().md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+`args.customClass` | Custom class for the label if needed | `String` | | `false`
```
Comment (azefiel) [inline docs/monster-ui/charsRemaining().md]:
```suggestion
`customClass` | Custom class for the label if needed | `String` | | `false`
```
(from [example](https://raw.githubusercontent.com/2600hz/monster-ui/master/docs/monster-ui/monthpicker().md))

File: src/js/lib/monster.ui.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (azefiel) [reply src/js/lib/monster.ui.js]: If that's the case, this behavior should then be described in the documentation.

**Comment by masmerino13** [comment]
Hi @azefiel 

Are you able to approve it again?

**Pull request merged**